docs/guides/pallas_kernels_performance.md

```diff
@@ -0,0 +1,127 @@
+<!--
+ Copyright 2023–2025 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+
+# Performance Optimizations with Pallas Kernels
+
+## Introduction
+
+While JAX and the XLA compiler provide excellent out-of-the-box performance, writing custom kernels with Pallas, a JAX extension for GPUs and TPUs, can unlock even greater speed. Pallas allows for lower-level control over hardware execution, which is especially beneficial for memory-bound operations. This guide provides an overview of writing and integrating custom Pallas kernels to optimize MaxText's performance.
+
+## Benefits of Using Pallas
+
+While the XLA compiler is highly effective, Pallas provides a powerful tool for expert users to unlock additional performance in specific scenarios. Key benefits include:
+
+* **Operator Fusion**: Manually fuse multiple operations into a single kernel to reduce launch overhead and memory I/O, especially when XLA's automatic fusion is suboptimal.
+* **Hardware-Specific Optimizations**: Tailor code to the specific architecture of your GPU or TPU, leveraging features that the compiler might not fully utilize. This includes fine-grained control over memory access and parallelism.
+* **Improved Memory Access Patterns**: With Pallas, you can optimize how data is read from and written to memory to maximize bandwidth and reduce latency.
+* **Algorithmic Optimizations**: Implement novel algorithms not available in standard libraries, such as flash attention, which combines the benefits of the above points to significantly speed up attention and save memory.
+
+### When Not to Use Pallas
+
+Pallas is a specialized tool and not always the right choice. Stick with standard JAX in these scenarios:
+
+* **When XLA is Sufficient**: If profiling shows your code is already efficient, a custom kernel adds complexity for little gain. **Always profile first.**
+* **For Purely Compute-Bound Operations**: Pallas excels at memory-bound problems. For large, compute-bound operations like `jnp.matmul`, it's hard to beat XLA's highly optimized library calls.
```

Comment:
Do we have a source of this statement? I am not fully convinced. In the MLP block of MoE model, the `jnp.matmul` is not efficient to handle ragged tensors. It could be compute bound, but some of compute are wasteful for non-selected experts. Either using a ragged version of matmul or pallas kernel could customize the needed compute for speed up.